src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Android/StorePal.LoaderPal.cs

```diff
@@ -13,29 +14,38 @@ internal sealed partial class StorePal
     {
         private sealed class AndroidCertLoader : ILoaderPal
         {
-            private readonly SafeX509Handle[] _certs;
+            private ICertificatePal[]? _certs;
 
-            public AndroidCertLoader(SafeX509Handle[] certs)
+            public AndroidCertLoader(SafeX509Handle[] certHandles)
+            {
+                _certs = new ICertificatePal[certHandles.Length];
+                for (int i = 0; i < certHandles.Length; i++)
+                {
+                    SafeX509Handle handle = certHandles[i];
+                    Debug.Assert(!handle.IsInvalid);
+                    _certs[i] = AndroidCertificatePal.FromHandle(handle.DangerousGetHandle());
+                }
+            }
+
+            public AndroidCertLoader(ICertificatePal[] certs)
             {
                 _certs = certs;
             }
 
             public void Dispose()
             {
-                foreach (var cert in _certs)
-                    cert.Dispose();
+                if (_certs != null)
+                    _certs.DisposeAll();
             }
 
             public void MoveTo(X509Certificate2Collection collection)
```

Comment:
What is this is called twice on the same instance? Do we have some higher level invariant here to avoid that error?

Reply:
I think we're just relying on the consumption in `X509Certificate2Collection` to only do it once.

I can make it actually return if it has already been moved instead of just asserting in debug.

Reply:
The assert is fine